Preview extraction fails on short videos

When ingesting a video with a length shorter than 5s, the extraction of a preview image will fail due to the configuration of the default workflow.

```
2020-01-05 02:12:18,761 | ERROR | (AbstractJobProducer$JobRunner:344) - Error handling operation 'Image': org.opencastproject.serviceregistry.api.ServiceRegistryException: Error handling operation 'Image'                                  
        at org.opencastproject.composer.impl.ComposerServiceImpl.process(ComposerServiceImpl.java:1530)                                                                                                                                       
        at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:314)                                                                                                                                       
        at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:273)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.opencastproject.composer.api.EncoderException: An image could not be extracted from the track http://localhost:8080/files/mediapackage/9bf975ce-67ad-406a-9193-9e0807c13d20/f7544c6d-69c2-4943-85e5-9b66432b0a5c/0_5s.mp4 with 
id f7544c6d-69c2-4943-85e5-9b66432b0a5c because the extraction time (1.0 second(s)) is outside of the track's duration (0.535 second(s))
        at org.opencastproject.composer.impl.ComposerServiceImpl.image(ComposerServiceImpl.java:1121)
        at org.opencastproject.composer.impl.ComposerServiceImpl.process(ComposerServiceImpl.java:1422)
        ... 6 more

```

--- a/etc/encoding/engage-images.properties
+++ b/etc/encoding/engage-images.properties
@@ -33,7 +33,9 @@ profile.player-preview.http.name = cover image for engage
 profile.player-preview.http.input = visual
 profile.player-preview.http.output = image
 profile.player-preview.http.suffix = -player.jpg
-profile.player-preview.http.ffmpeg.command = -ss #{time} -i #{in.video.path} -r 1 -frames:v 1 -filter:v yadif,scale=-1:480 #{out.dir}/#{out.name}#{out.suffix}
+profile.player-preview.http.ffmpeg.command = -i #{in.video.path} -frames:v 1 \
+  -filter:v trim=0:#{time},fps=1,reverse,scale=-2:480 \
+  #{out.dir}/#{out.name}#{out.suffix}
 
 # Slide preview images as shown in the player
 profile.player-slides.http.name = slide preview image for engage
@@ -47,11 +49,12 @@ profile.search-cover.http.name = cover image for engage
 profile.search-cover.http.input = visual
 profile.search-cover.http.output = image
 profile.search-cover.http.suffix = -search.jpg
-profile.search-cover.http.ffmpeg.command = -ss #{time} -i #{in.video.path} -r 1 -frames:v 1 -filter:v yadif,scale=160:-1 #{out.dir}/#{out.name}#{out.suffix}
+profile.search-cover.http.ffmpeg.command = -i #{in.video.path} -frames:v 1 \
+  -filter:v trim=0:#{time},fps=1,reverse,scale=160:-2 \
+  #{out.dir}/#{out.name}#{out.suffix}
 
 profile.search-cover.http.downscale.name = Downscale to cover image for engage
 profile.search-cover.http.downscale.input = visual
 profile.search-cover.http.downscale.output = image
 profile.search-cover.http.downscale.suffix = -search.jpg
 profile.search-cover.http.downscale.ffmpeg.command = -i #{in.video.path} -r 1 -frames:v 1 -filter:v yadif,scale=160:-1 #{out.dir}/#{out.name}#{out.suffix}
-

--- a/modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java
+++ b/modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java
@@ -100,7 +100,6 @@ import java.net.URI;
 import java.net.URISyntaxException;
 import java.text.DecimalFormat;
 import java.text.DecimalFormatSymbols;
-import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -1042,7 +1041,7 @@ public class ComposerServiceImpl extends AbstractJobProducer implements Composer
               JOB_TYPE, Operation.Image.toString(), null, null, false, profile.getJobLoad());
       job.setStatus(Job.Status.RUNNING);
       job = serviceRegistry.updateJob(job);
-      final List<Attachment> images = image(job, sourceTrack, profileId, time);
+      final List<Attachment> images = extractImages(job, sourceTrack, profileId, null, time);
       job.setStatus(Job.Status.FINISHED);
       return images;
     } catch (ServiceRegistryException | NotFoundException e) {
@@ -1081,78 +1080,19 @@ public class ComposerServiceImpl extends AbstractJobProducer implements Composer
    *          the source track
    * @param profileId
    *          the identifier of the encoding profile to use
+   * @param properties
+   *          the properties applied to the encoding profile
    * @param times
    *          (one or more) times in seconds
    * @return the images as an attachment element list
    * @throws EncoderException
    *           if extracting the image fails
    */
-  protected List<Attachment> image(Job job, Track sourceTrack, String profileId, double... times)
-          throws EncoderException, MediaPackageException {
-    if (sourceTrack == null)
-      throw new EncoderException("SourceTrack cannot be null");
-
-    validateVideoStream(job, sourceTrack);
-
-    // The time should not be outside of the track's duration
-    for (double time : times) {
-      if (sourceTrack.getDuration() == null) {
-        Map<String, String> params = new HashMap<>();
-        params.put("track-id", sourceTrack.getIdentifier());
-        params.put("track-url", sourceTrack.getURI().toString());
-        incident().recordFailure(job, IMAGE_EXTRACTION_UNKNOWN_DURATION, params);
-        throw new EncoderException("Unable to extract an image from a track with unknown duration");
-      }
-      if (time < 0 || time * 1000 > sourceTrack.getDuration()) {
-        Duration duration = Duration.ofMillis(sourceTrack.getDuration());
-        long durationSeconds = duration.getSeconds();
-        long durationMillis = duration.minusSeconds(durationSeconds).toMillis();
-        String formattedDuration = String.valueOf(durationSeconds) + '.'
-                + StringUtils.leftPad(String.valueOf(durationMillis), 3, '0');
-
-        Map<String, String> params = new HashMap<>();
-        params.put("track-id", sourceTrack.getIdentifier());
-        params.put("track-url", sourceTrack.getURI().toString());
-        params.put("track-duration", formattedDuration);
-        params.put("time", Double.toString(time));
-        incident().recordFailure(job, IMAGE_EXTRACTION_TIME_OUTSIDE_DURATION, params);
-
-        throw new EncoderException("An image could not be extracted from the track " + sourceTrack.getURI()
-                + " with id " + sourceTrack.getIdentifier() + " because the extraction time (" + time + " second(s)) is "
-                + "outside of the track's duration (" + formattedDuration + " second(s))");
-      }
-    }
-
-    return extractImages(job, sourceTrack, profileId, null, times);
-  }
-
-  /**
-   * Extracts an image from <code>sourceTrack</code> by the given properties and the corresponding encoding profile.
-   *
-   * @param job
-   *          the associated job
-   * @param sourceTrack
-   *          the source track
-   * @param profileId
-   *          the identifier of the encoding profile to use
-   * @param properties
-   *          the properties applied to the encoding profile
-   * @return the images as an attachment element list
-   * @throws EncoderException
-   *           if extracting the image fails
-   */
-  protected List<Attachment> image(Job job, Track sourceTrack, String profileId, Map<String, String> properties)
-          throws EncoderException, MediaPackageException {
-    if (sourceTrack == null)
-      throw new EncoderException("SourceTrack cannot be null");
-
-    validateVideoStream(job, sourceTrack);
-
-    return extractImages(job, sourceTrack, profileId, properties);
-  }
-
   private List<Attachment> extractImages(Job job, Track sourceTrack, String profileId, Map<String, String> properties,
           double... times) throws EncoderException {
+    if (sourceTrack == null) {
+      throw new EncoderException("SourceTrack cannot be null");
+    }
     logger.info("creating an image using video track {}", sourceTrack.getIdentifier());
 
     // Get the encoding profile
@@ -1419,10 +1359,10 @@ public class ComposerServiceImpl extends AbstractJobProducer implements Composer
             for (int i = 3; i < arguments.size(); i++) {
               times[i - 3] = Double.parseDouble(arguments.get(i));
             }
-            resultingElements = image(job, firstTrack, encodingProfile, times);
+            resultingElements = extractImages(job, firstTrack, encodingProfile, null, times);
           } else {
             Map<String, String> properties = parseProperties(arguments.get(3));
-            resultingElements = image(job, firstTrack, encodingProfile, properties);
+            resultingElements = extractImages(job, firstTrack, encodingProfile, properties);
           }
           serialized = MediaPackageElementParser.getArrayAsXml(resultingElements);
           break;

